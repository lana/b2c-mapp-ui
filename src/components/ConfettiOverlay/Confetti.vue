<template>
  <div ref="confettiRef" class="confetti-container" :style="containerStyle">
    <div class="confetti" :class="color" :style="confettiStyle"/>
  </div>
</template>

<script>
import { ref, onMounted } from 'vue';

export default {
  props: {
    color: {
      type: String,
      default: 'red',
    },
    width: {
      type: Number,
      default: 10,
    },
    height: {
      type: Number,
      default: 10,
    },
  },
  setup(props) {
    const confettiRef = ref(null);
    const containerStyle = {
      visibility: 'hidden',
      width: `${props.width}px`,
      height: `${props.width * 0.8}px`,
      left: `${props.height}%`,
      opacity: Math.random() + 0.5,
    };
    const confettiStyle = {
      transform: `rotate(${Math.random() * 360}deg)`,
    };
    onMounted(() => {
      const plusOrMinus = (Math.random() < 0.5) ? (-1) : (1);
      const animation = [
        {
          visibility: 'visible',
        },
        {
          top: '110%',
          transform: `translateX(${plusOrMinus * props.height * Math.random() * 15}%)`,
        }];
      const animationProps = { duration: (2 + Math.random()) * 1000, delay: Math.random() * 1000, easing: 'ease-in' };
      confettiRef.value.animate(animation, animationProps);
    });
    return { containerStyle, confettiStyle, confettiRef };
  },
};
</script>
