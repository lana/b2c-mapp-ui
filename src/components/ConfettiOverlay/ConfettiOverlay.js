const defaultCountOfParticles = 80;
const maximumCountOfParticlesForSmoothPerformance = 200;

const props = {
  particles: {
    type: [Number, String],
    default: defaultCountOfParticles,
    validator: (value) => {
      const numericValue = Number.parseInt(value, 10);
      const isValid = !Number.isNaN(numericValue);
      return (isValid && numericValue <= maximumCountOfParticlesForSmoothPerformance);
    },
  },
};

const computed = {
  countOfParticles() {
    const numericValue = Number.parseInt(this.particles, 10);
    const result = (numericValue < maximumCountOfParticlesForSmoothPerformance) ? numericValue : maximumCountOfParticlesForSmoothPerformance;
    return result;
  },
};

const ConfettiOverlay = {
  props,
  computed,
};

export default ConfettiOverlay;
