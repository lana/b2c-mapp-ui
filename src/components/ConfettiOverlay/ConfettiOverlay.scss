@import '../../styles/colors';

$colors: "blue" $blue-600, "orange" $orange-500, "pink" $pink-600, "purple" $purple-400, "red" $red-500, "yellow" $yellow-500;

.wrapper {
  position: absolute;
  height: 100%;
  width: 100%;
  margin: 0;
  padding:0;
  overflow: hidden;
}

.confetti-container {
  position: absolute;
  top: -10%;
}

:deep(.confetti) {
  width: 100%;
  height: 100%;

  @each $colorName, $color in $colors {
    &.#{$colorName} {
      background-color: $color;
    }
  }
}
