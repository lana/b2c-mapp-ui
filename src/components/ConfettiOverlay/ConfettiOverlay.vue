<template>
  <div class="wrapper">
    <div v-for="index in countOfParticles"
         :key="index"
         :class="`confetti-${index}`"
    />
  </div>
</template>
<script src="./ConfettiOverlay.js"/>
<style lang="scss" scoped src="./ConfettiOverlay.scss"/>
