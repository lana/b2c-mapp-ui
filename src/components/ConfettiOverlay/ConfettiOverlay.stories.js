import ConfettiOverlay from './ConfettiOverlay.vue';
import ScrollWrapper from '../ScrollWrapper/ScrollWrapper.vue';
import Screen from '../Screen/Screen.vue';
import Heading from '../Heading/Heading.vue';
import Button from '../Button/Button.vue';
import { createDeviceDecorator } from '../../lib/storybookHelpers';

const deviceDecorator = createDeviceDecorator('<strong>ConfettiOverlay:</strong>');

const ConfettiStories = {
  component: ConfettiOverlay,
  title: 'Components/ConfettiOverlay',
  decorators: [deviceDecorator],
  args: {
    ...deviceDecorator.args,
    particles: 50,
  },
  argTypes: {
    ...deviceDecorator.argTypes,
    particles: { name: 'Count of Particles', control: 'number' },
  },
};

const defaultExample = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: {
    ConfettiOverlay,
    Button,
    Heading,
    Screen,
    ScrollWrapper,
  },
  data() {
    return {
      isConfettiShowing: true,
    };
  },
  methods: {
    async restartConfetti() {
      this.isConfettiShowing = false;
      await this.$nextTick();
      this.isConfettiShowing = true;
    },
  },
  template: `
    <Screen>
      <ScrollWrapper>
        <ConfettiOverlay v-if="isConfettiShowing" :particles="particles"/>
        <Button style="margin-bottom: 10px;" @click="restartConfetti">
          <p>Restart Confetti</p>
        </Button>
        <Heading style="text-align: center;"
                size="xxxl"
                weight="semibold"
        >
          DEAL WITH IT
        </Heading>
      </ScrollWrapper>
      <img style="margin: auto; width: 200px; height: auto;" src="https://img.pngio.com/dancing-gifs-get-the-best-gif-on-giphy-dancing-people-png-gif-350_497.gif">
    </Screen>
   `,
});
defaultExample.parameters = {
  docs: {
    source: {
      code: `
<Screen>
  <ScrollWrapper>
    <ConfettiOverlay v-if="isConfettiShowing" :particles="particles"/>
  </ScrollWrapper>
</Screen>`,
    },
  },
};

export {
  defaultExample,
};

export default ConfettiStories;
