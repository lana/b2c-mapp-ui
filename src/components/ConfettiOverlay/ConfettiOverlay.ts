import { defineComponent } from 'vue';

import Confetti from './Confetti.vue';

const defaultCountOfParticles = 80;
const maximumCountOfParticlesForSmoothPerformance = 200;
const colors = ['blue', 'orange', 'pink', 'purple', 'red', 'yellow'];

const ConfettiOverlay = defineComponent({
  name: 'ConfettiOverlay',
  components: {
    Confetti,
  },
  props: {
    particles: {
      type: [Number, String],
      default: defaultCountOfParticles,
      validator: (value: number | string) => {
        const numericValue = Number.parseInt(`${value}`, 10);
        const isValid = !Number.isNaN(numericValue);
        const result = (isValid && (numericValue <= maximumCountOfParticlesForSmoothPerformance));
        return result;
      },
    },
  },
  computed: {
    countOfParticles() {
      const numericValue = Number.parseInt(`${this.particles}`, 10);
      const count = (numericValue < maximumCountOfParticlesForSmoothPerformance) ? numericValue : maximumCountOfParticlesForSmoothPerformance;
      const result = Array(count).fill(0).map((_, index) => ({
        id: index,
        color: colors[Math.floor(Math.random() * colors.length)],
        width: Math.random() * 10,
        height: Math.random() * 100,
      }));
      return result;
    },
  },
});

export default ConfettiOverlay;
