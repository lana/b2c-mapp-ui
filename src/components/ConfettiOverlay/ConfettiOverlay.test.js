import { shallowMount } from '@vue/test-utils';

import ConfettiOverlay from './ConfettiOverlay.vue';

describe('ConfettiOverlay unit test', () => {
  const waitForDomUpdate = async (wrapper) => {
    await wrapper.vm.$nextTick();
    await wrapper.vm.$forceUpdate();
  };

  it('Should pass health-check test', async () => {
    const wrapper = shallowMount(ConfettiOverlay, {});
    await waitForDomUpdate(wrapper);
    expect(wrapper.findAll('.wrapper > *').length).toBe(80);
  });
});
