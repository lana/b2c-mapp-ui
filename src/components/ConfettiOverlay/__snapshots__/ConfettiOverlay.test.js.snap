// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfettiOverlay unit test Should pass health-check test 1`] = `
<div class="wrapper">
  <div class="confetti-1"></div>
  <div class="confetti-2"></div>
  <div class="confetti-3"></div>
  <div class="confetti-4"></div>
  <div class="confetti-5"></div>
  <div class="confetti-6"></div>
  <div class="confetti-7"></div>
  <div class="confetti-8"></div>
  <div class="confetti-9"></div>
  <div class="confetti-10"></div>
  <div class="confetti-11"></div>
  <div class="confetti-12"></div>
  <div class="confetti-13"></div>
  <div class="confetti-14"></div>
  <div class="confetti-15"></div>
  <div class="confetti-16"></div>
  <div class="confetti-17"></div>
  <div class="confetti-18"></div>
  <div class="confetti-19"></div>
  <div class="confetti-20"></div>
  <div class="confetti-21"></div>
  <div class="confetti-22"></div>
  <div class="confetti-23"></div>
  <div class="confetti-24"></div>
  <div class="confetti-25"></div>
  <div class="confetti-26"></div>
  <div class="confetti-27"></div>
  <div class="confetti-28"></div>
  <div class="confetti-29"></div>
  <div class="confetti-30"></div>
  <div class="confetti-31"></div>
  <div class="confetti-32"></div>
  <div class="confetti-33"></div>
  <div class="confetti-34"></div>
  <div class="confetti-35"></div>
  <div class="confetti-36"></div>
  <div class="confetti-37"></div>
  <div class="confetti-38"></div>
  <div class="confetti-39"></div>
  <div class="confetti-40"></div>
  <div class="confetti-41"></div>
  <div class="confetti-42"></div>
  <div class="confetti-43"></div>
  <div class="confetti-44"></div>
  <div class="confetti-45"></div>
  <div class="confetti-46"></div>
  <div class="confetti-47"></div>
  <div class="confetti-48"></div>
  <div class="confetti-49"></div>
  <div class="confetti-50"></div>
  <div class="confetti-51"></div>
  <div class="confetti-52"></div>
  <div class="confetti-53"></div>
  <div class="confetti-54"></div>
  <div class="confetti-55"></div>
  <div class="confetti-56"></div>
  <div class="confetti-57"></div>
  <div class="confetti-58"></div>
  <div class="confetti-59"></div>
  <div class="confetti-60"></div>
  <div class="confetti-61"></div>
  <div class="confetti-62"></div>
  <div class="confetti-63"></div>
  <div class="confetti-64"></div>
  <div class="confetti-65"></div>
  <div class="confetti-66"></div>
  <div class="confetti-67"></div>
  <div class="confetti-68"></div>
  <div class="confetti-69"></div>
  <div class="confetti-70"></div>
  <div class="confetti-71"></div>
  <div class="confetti-72"></div>
  <div class="confetti-73"></div>
  <div class="confetti-74"></div>
  <div class="confetti-75"></div>
  <div class="confetti-76"></div>
  <div class="confetti-77"></div>
  <div class="confetti-78"></div>
  <div class="confetti-79"></div>
  <div class="confetti-80"></div>
</div>
`;
